File: 367.cpp
class Solution {
public:
    bool isPerfectSquare(int num) {
        int l = 1, r = num;
        while(l<=r){
            int m = (l+r)/2;
            if(num%m==0 && m==num/m){
                return true;
            }else if(m<=num/m){
                l = m+1;
            }else{
                r = m-1;
            }
        }
        return false;
    }
};

File: go-solutions/367.go
package main

func isPerfectSquare(num int) bool {
	if num == 0 {
		return false
	}

	left, right := 1, num
	for left <= right {
		mid := left + (right-left)/2
		switch {
		case mid*mid == num:
			return true

		case mid*mid < num:
			left = mid + 1

		default:
			right = mid - 1
		}
	}
	return right*right == num
}
